pkg/controllers/namespace/namespace_sync_controller_test.go

```diff
@@ -0,0 +1,505 @@
+	}, work)
+	assert.NoError(t, err)
+	assert.NotNil(t, work)
+
+	t.Logf("Work found: %+v", work)
+	t.Logf("Work annotations: %v", work.Annotations)
+	t.Logf("Work spec: %+v", work.Spec)
+
+	assert.NotNil(t, work)
+}
+
+func TestController_SetupWithManager(t *testing.T) {
+	scheme := runtime.NewScheme()
+	err := corev1.AddToScheme(scheme)
+	assert.NoError(t, err)
+	err = clusterv1alpha1.Install(scheme)
+	assert.NoError(t, err)
+	err = workv1alpha1.Install(scheme)
+	assert.NoError(t, err)
+	err = policyv1alpha1.Install(scheme)
+	assert.NoError(t, err)
+
+	cfg := &rest.Config{Host: "http://localhost:8080"}
+	mgr, err := controllerruntime.NewManager(cfg, controllerruntime.Options{
+		Scheme: scheme,
+	})
+	assert.NoError(t, err)
+
+	c := &Controller{
+		Client:          mgr.GetClient(),
+		EventRecorder:   mgr.GetEventRecorderFor("test-controller"),
+		OverrideManager: overridemanager.New(mgr.GetClient(), mgr.GetEventRecorderFor("test-controller")),
+	}
+
+	err = c.SetupWithManager(mgr)
+	assert.NoError(t, err)
+
+	assert.NotNil(t, c.Client, "Controller's Client should not be nil")
+	assert.NotNil(t, c.EventRecorder, "Controller's EventRecorder should not be nil")
+	assert.NotNil(t, c.OverrideManager, "Controller's OverrideManager should not be nil")
+}
+
+func TestController_SetupWithManagerConcurrent(t *testing.T) {
+	scheme := runtime.NewScheme()
+	err := corev1.AddToScheme(scheme)
+	assert.NoError(t, err)
+	err = clusterv1alpha1.Install(scheme)
+	assert.NoError(t, err)
+	err = workv1alpha1.Install(scheme)
+	assert.NoError(t, err)
+	err = policyv1alpha1.Install(scheme)
+	assert.NoError(t, err)
+
+	cfg := &rest.Config{Host: "http://localhost:8080"}
+	mgr, err := controllerruntime.NewManager(cfg, controllerruntime.Options{
+		Scheme: scheme,
+	})
+	assert.NoError(t, err)
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			c := &Controller{
+				Client:          mgr.GetClient(),
+				EventRecorder:   mgr.GetEventRecorderFor("test-controller"),
+				OverrideManager: overridemanager.New(mgr.GetClient(), mgr.GetEventRecorderFor("test-controller")),
+			}
+			err := c.SetupWithManager(mgr)
+			assert.NoError(t, err)
+		}()
+	}
+	wg.Wait()
+}
+
+func TestController_SetupWithManagerError(t *testing.T) {
+	scheme := runtime.NewScheme()
+	// Schemes are not added to intentionaly trigger an error
+
+	cfg := &rest.Config{Host: "http://localhost:8080"}
+	mgr, err := controllerruntime.NewManager(cfg, controllerruntime.Options{
+		Scheme: scheme,
+	})
+	assert.NoError(t, err)
+
+	c := &Controller{
+		Client:          mgr.GetClient(),
+		EventRecorder:   mgr.GetEventRecorderFor("test-controller"),
+		OverrideManager: overridemanager.New(mgr.GetClient(), mgr.GetEventRecorderFor("test-controller")),
+	}
+
+	err = c.SetupWithManager(mgr)
+	assert.Error(t, err, "SetupWithManager should return an error when schemes are not properly set up")
+}
+
+func TestClusterNamespaceFn(t *testing.T) {
```

Comment:
Sorry, I didn't understand which function body the function tests.

Reply:
The two you mentioned are clusterNamespaceFn and clusterOverridePolicyNamespaceFn which are MapFuncs used in the controller's SetupWithManager() method to handle events related to Cluster and ClusterOverridePolicy resources.

Reply:
I understand what you are saying, but where is the code guarded by these two test methods?

Reply:
Okay, i got it. I think i need to refractor the whole  controller struct. Thank You for pointing out.

Reply:
@XiShanYongYe-Chang I'm not able to test them in isolation. I think we should ahead without them.

Reply:
Yes, I don't think they need to be tested.